### ISSUE TITLE
[Documentation] Protected web APIs that call downstream APIs on behalf of a user

### ISSUE THREAD

**Issue by arqe** [AzureAD/microsoft-identity-web#915]
### Documentation related to component
Protected web APIs that call downstream APIs on behalf of a user

### Please check all that apply

- [ ] typo
- [ ] documentation doesn't exist
- [x] documentation needs clarification
- [x] error(s) in the example
- [ ] needs an example

### Description of the issue

Hi, 
I am having issues with the section [https://github.com/AzureAD/microsoft-identity-web/wiki/web-apis#azure-ad-b2c](https://github.com/AzureAD/microsoft-identity-web/wiki/web-apis#azure-ad-b2c)

My understanding is that this is the OBO flow, which is (quite sadly!) not supported by B2C. Is my understanding correct?

Before finding out about the B2C OBO limitation, I tried to used the same construction as the one documented in the linked section: 

**API secured with B2C**

```using Microsoft.Identity.Web;
public class Startup
{
  ...
  public void ConfigureServices(IServiceCollection services)
  {
   ...
   services.AddMicrosoftIdentityWebApiAuthentication(Configuration, "AzureAdB2C")
             .EnableTokenAcquisitionToCallDownstreamApi()
                .AddInMemoryTokenCaches();
   ...
  }
  ...
}
```

**Its Controller calling downstream API**

```[Authorize]
public class HomeController : Controller
{
  readonly ITokenAcquisition tokenAcquisition;

  static string[] scopeRequiredByAPI = new string[] { "access_as_user" };
  ...
  public async Task<IActionResult> Action()
  {
   HttpContext.VerifyUserHasAnyAcceptedScope(scopeRequiredByAPI);
   string[] scopes = new []{"user.read"};
   try
   {
      string accessToken = await _tokenAcquisition.GetAccessTokenOnBehalfOfUser(scopes);
      // call the downstream API with the bearer token in the Authorize header
    }
    catch (MsalUiRequiredException ex)
    {
      await _tokenAcquisition.ReplyForbiddenWithWwwAuthenticateHeaderAsync(HttpContext, scopes, ex);
    }
   ...
}
```

The API is accessed on behalf with scope "**access_as_user**" and tries to get access token for downstream API by **_tokenAcquisition.GetAccessTokenOnBehalfOfUser(scopes);** I believe that one was meant to actually be: **_tokenAcquisition.GetAccessTokenForUserAsync(scopes);** , as referred in the document section above the code snippet: 

> In your controller actions, call **ITokenAcquisition.GetAccessTokenForUserAsync**, passing the scopes for which to request a token.

Nevertheless, this approach failed on me consistently with:
```
MSAL.NetCore.4.25.0.0.MsalServiceException: 
	ErrorCode: unsupported_grant_type
Microsoft.Identity.Client.MsalServiceException: AADB2C90086: The supplied grant_type [urn:ietf:params:oauth:grant-type:jwt-bearer] is not supported.
```

Which I now believe is legit, because of [https://docs.microsoft.com/en-us/azure/active-directory-b2c/access-tokens](url): 

> Web API chains (On-Behalf-Of) is not supported by Azure AD B2C. - Many architectures include a web API that needs to call another downstream web API, both secured by Azure AD B2C. This scenario is common in clients that have a web API back end, which in turn calls a another service. This chained web API scenario can be supported by using the OAuth 2.0 JWT Bearer Credential grant, otherwise known as the On-Behalf-Of flow. However, the On-Behalf-Of flow is not currently implemented in Azure AD B2C.

This leads me to a conclusion, that the documentation is not correct, and that using B2C for this OBO flow is not possible at the moment. Quite sadly, again, if I am correct (I still hope I am missing something).

**Comment by jmprieur** [comment]
Thanks for the heads-up, @arge, and for your very clear and actionable report.
you are right the documentation was incorrect.

Which sample are you referring to? I don't think that we have a B2C web api sample that calls a downstream API. Would it be the code snippet in the page?

- I just fixed the doc (and actually refactored a lot the [web api])(https://github.com/AzureAD/microsoft-identity-web/wiki/web-apis#azure-ad-b2c) page.
- I also updated the [Azure AD B2C limitations](b2c-limitations) page with the link to the B2C documentation you provided (and which I didn't know about)

You're right that this is sad that OBO is not supported. @jennyf19 mentioned that there might be an ugly work around using AAD apps. If that's the case, we'll document it and we'll keep you posted

**Comment by arqe** [comment]
@jmprieur Thanks a lot! I believe this update will be very useful for many others. 

I was only referring to the code snippets in the documentation, the "error(s) in the example" was wrongly ticked.  

Also, knowing workaround would be absolutely great! 

Many thanks, 

ARQE

**Comment by jennyf19** [comment]
Included in [1.6.0](https://github.com/AzureAD/microsoft-identity-web/releases/tag/1.6.0) release.
